### ISSUE TITLE
latex figcaptions

### ISSUE THREAD

**Issue by olehwaa** [jfbercher/jupyter_latex_envs#16]
Hi,
Should figure captions generated with latex_envs.figcaption become figure captions in latex document?
I didn't understand how so I modified your code to satisfy my needs. It also works with multiple figures in a output cell.
The idea is:
- Create a JSON field with the figcaption function:
```python
def figcaption(text, label=" "):
    display(HTML("<div class=caption><b> Caption: </b> %s</div>"
                 % text.replace('\n', '<br>')))
    text = text.replace('<b>',r'\textbf{').replace('</b>','}') #some replacement of HTML formatting
    text = text.replace('<i>',r'\textit{').replace('</i>','}')
    text = text.replace('<tt>',r'\texttt{').replace('</tt>','}')
    display(JSON({'caption': text, 'label':label}),include=["application/json"])
```
- In LenvsLatexPreprocessor.preprocess_cell function populate the output cells metadata with a figure with adjacent JSON field:
```python
if cell.cell_type == "code" and "outputs" in cell:
    json_metadata = []
    mime_types  = ['image/svg+xml','image/png','image/jpeg','application/pdf']
    for output in cell.outputs:
        if "data" in output:
            if  'application/json' in output.data and 'caption' in output.data['application/json']: #found a json field with caption
                json_metadata.append(output.data['application/json'])
                
            if any(x in output.data for x in mime_types) and len(json_metadata)>0:
                output.metadata.update(json_metadata.pop()) # write caption to data field metadata
```
- Use the template structure described in [Making publication ready Python Notebooks](http://blog.juliusschulz.de/blog/ultimate-ipython-notebook#figures-1)
Replace cell.metadata.caption with output.metadata.caption and  cell.metadata.label with output.metadata.label in this template

Hope you find this way to generate figure captions interesting...

**Comment by jfbercher** [comment]
This is certainly interesting, thanks!

What I do not understand is that export of figures captions is already implemented. It seems that it doesn't work for you? If so, can you explain me what happens? 
The following snippet is an example:
```
import numpy as np
import matplotlib.pyplot as plt 
from latex_envs.latex_envs import figcaption 
%matplotlib inline
plt.plot(np.random.randn((100)))
figcaption("This is the caption", label="fig:attempt")
```
Then issue 
`jupyter nbconvert --to latex_with_lenvs FILE.ipynb `
This is explained [here](https://rawgit.com/jfbercher/jupyter_latex_envs/master/src/latex_envs/static/doc/latex_env_doc.html#figcaption-27) and [here](https://rawgit.com/jfbercher/jupyter_latex_envs/master/src/latex_envs/static/doc/latex_env_doc.html#Conversion-to-LaTeX-32) in the documentation. 

Anyway, your way of doing looks more interesting and systematic than the ad hoc search and replace I  do in the code, so I will try to include it in a next version. Don't you want to make a pull request yourself?

**Comment by jfbercher** [comment]
@olehwaa

**Comment by olehwaa** [comment]
Great that you find my suggestions interesting!

I admit I didn't test the figcaption functionality in it simplest form, so it might work.
What I wanted to do was to adapt my custom template based on
[Making publication ready Python Notebooks](http://blog.juliusschulz.de/blog/ultimate-ipython-notebook).
In this template I'm producing a document without the code cells. If I have understood this correctly, the word figcaption you are searching for is the figcaption comand in the code cell. Obviously, there will be no figure captions in the latex document if I remove to code cell....
Another problem is that this approch will not work if you programically set the captions, for instance if you produce plots in a loop and you alter the figure captions for each plot.
My solution has been tested to  handle these cases, so I think it's more versatile. 

The implementation of this concept should be discussed. Writing to the notebook file with display(JSON(d)) is a possibility since it is not actually displayed in HTML or Latex. However, someone might write an extension that display JSON cells, and this data is not meant for displaying. The preprocessing script copy this info into the  data subcell  metadata, and ideally the captions were writting into metadata of the  data subcells with figures, but I don't know if this is possible to do at all, especially if figcaption is called before the figure is produced.

Another option is to write to the cell metadata as a list of captions, and copy this into metadata of the  data subcells. I don't know how to do this, but I see that at least it is possible [to write to Notebook metadata](https://github.com/damianavila/RISE#configure-your-own-options) with traitlets.config.manager.BaseJSONConfigManager

Another thing I would like to have is references to figure set with label in function figcaption working in the notebook.

I'll make a poll request in a few days. (Since this is the first project I participate in on GitHub, I have learn how to make poll requests...)

**Comment by jfbercher** [comment]
Hi @olehwaa 
I just updated latex_envs to add some autocompletion capabilities. I understand your usecase and an update will be welcome. I was thinking that it is probably possible to directly write thinks to the cell metadata using a trick like
`Javascript("var c = IPython.notebook.get_selected_cell(); c.metadata.test = false; console.log(c.metadata)")` (where I create a test key with a "false" value)
For doing a pull request, you can proceed as follows : 
- make a fork of this repo
- clone the fork to your disk
- work on these files. Each time you update something, or at each important step do
     - add files with `git add .`
     - commit with `git commit -m "descriptive_message_of_what_was_done"`
- when you have finish or are ready, push the modified files to your repo by `git push`
- go to your repo on github web and a "make pull request button" should pop up and then you will just have to follow the instructions.

**Comment by jfbercher** [comment]
Closed by #17
